dask_mongo/core.py

```diff
@@ -14,19 +18,68 @@
 
 appname = f"dask-mongo/{__version__}"
 
+_CACHE_SIZE = 10
+_CLIENTS = weakref.WeakValueDictionary({})
+
+
+def _recursive_tupling(item):
+    if isinstance(item, list):
+        return tuple([_recursive_tupling(i) for i in item])
+    if isinstance(item, Mapping):
+        return tuple(
+            [(_recursive_tupling(k), _recursive_tupling(v)) for k, v in item.items()]
+        )
+    elif isinstance(item, Hashable):
+        return hash(item)
+    else:
+        return item
+
+
+class HashableKwargs(dict):
+    def __hash__(self):
+        return hash(
+            frozenset(
+                [
+                    (_recursive_tupling(k), _recursive_tupling(v))
+                    for k, v in self.items()
+                ]
+            )
+        )
+
+
+@lru_cache(_CACHE_SIZE, typed=True)
+def _cache_inner(kwargs):
+    return pymongo.MongoClient(appname=appname, **kwargs)
+
+
+@atexit.register
+def _close_clients():
+    global _CLIENTS
+    for func in _CLIENTS.values():
+        ref = func()
+        if ref:
+            ref()
+
+
+def _get_client(kwargs):
+    frozen_kwargs = HashableKwargs(kwargs)
+    client = _cache_inner(frozen_kwargs)
+    _CLIENTS[frozen_kwargs] = weakref.WeakMethod(client.close)
+    return client
```

Comment:
WeakValueDictionary is not being used correctly. In this case the dict will immediately remove the item that was added and it will always be empty. See https://docs.python.org/3/library/weakref.html